h2o-algos/src/main/java/hex/glm/ConstrainedGLMUtils.java

```diff
@@ -0,0 +1,861 @@
+      return true;
+    return false;
+  }
+  
+  public static boolean activeConstraints(LinearConstraints[] equalityC, LinearConstraints[] lessThanEqualToC) {
+    if (equalityC != null)
+      return true;
+    return stream(lessThanEqualToC).filter(x -> x._active).count() > 0;
+  }
+
+  /***
+   * This method calls getGradient to calculate the gradient, likelhood and objective function values.  In addition,
+   * it will add to the gradient and objective function the contribution from the linear constraints.
+   */
+  public static GLM.GLMGradientInfo calGradient(double[] betaCnd, ComputationState state, GLM.GLMGradientSolver ginfo,
+                                                double[] lambdaE, double[] lambdaL, LinearConstraints[] constraintE,
+                                                LinearConstraints[] constraintL) {
+    // todo: need to add support for predictors removed for whatever reason
+    // calculate gradients
+    GLM.GLMGradientInfo gradientInfo = ginfo.getGradient(betaCnd, state); // gradient without constraints
+    boolean hasEqualConstraints = constraintE != null;
+    boolean hasLessConstraints = constraintL != null;
+    List<String> coeffNames = Arrays.stream(state.activeData()._coefNames).collect(Collectors.toList());
+    // add gradient, objective and likelihood contribution from constraints
+    if (hasEqualConstraints) {
+      addConstraintGradient(lambdaE, state._derivativeEqual, gradientInfo);
+      addPenaltyGradient(state._derivativeEqual, constraintE, gradientInfo, state._csGLMState._ckCS);
+      gradientInfo._objVal += state.addConstraintObj(lambdaE, constraintE, betaCnd, state._csGLMState._ckCS, 
+              true, coeffNames);
+      //state.addConstraintObj(lambdaE, constraintE, state._csGLMState._ckCS, gradientInfo);
+    }
+    if (hasLessConstraints) {
+      addConstraintGradient(lambdaL, state._derivativeLess, gradientInfo);
+      addPenaltyGradient(state._derivativeLess, constraintL, gradientInfo, state._csGLMState._ckCS);
+      gradientInfo._objVal += state.addConstraintObj(lambdaL, constraintL, betaCnd, state._csGLMState._ckCS, 
+              false, coeffNames);
+    }
+    return gradientInfo;
+  }
+  
+  public static void updateConstraintValues(double[] betaCnd, List<String> coeffNames, 
+                                            LinearConstraints[] equalityConstraints, 
+                                            LinearConstraints[] lessThanEqualToConstraints) {
+    if (equalityConstraints != null) { // initialize lambda for equality constraints
+      Arrays.stream(equalityConstraints).collect(Collectors.toList()).parallelStream().forEach(constraint -> {
+        evalOneConstraint(constraint, betaCnd, coeffNames);
+        constraint._active = (Math.abs(constraint._constraintsVal) > EPS2);
+      });
+    }
+    if (lessThanEqualToConstraints != null)
+      Arrays.stream(lessThanEqualToConstraints).collect(Collectors.toList()).parallelStream().forEach(constraint -> {
```

Comment:
Please see https://github.com/h2oai/h2o-3/pull/16035/files#r1541134757 regarding the usage of parallelStream.